### PR TITLE
Fix username validation message

### PR DESCRIPTION
**Pull request by chen-robert**
Fixes part of https://github.com/publiclab/plots2/issues/6303

I fixed the broken authlogic regex for one character usernames [as described in this comment](https://github.com/publiclab/plots2/pull/6932#issuecomment-563878155). 

In addition, I added an integration test for the username validation errors returned from the /register endpoint.

### CODE REVIEW

**Comment by chen-robert** [comment]
@SidharthBansal when you have time, could you take a look please?

**Comment by chen-robert** [comment]
@Uzay-G The reason I refactored `@user.save` out into a new variable was because relying on the order of boolean statements for code correctness seems a bit error-prone. If a future contributor were to change the code, they might not realize the importance of the order of the two boolean statements. It's safer to leave it in a separate variable, which I think better communicates the intent of the code. 

Thanks for the help on the CI tests, hopefully we can get that fixed soon 🎉

**Comment by chen-robert** [comment]
@Uzay-G Thanks for the suggestion 😄 

I was actually going to do that initially, but the errors are already populated by authlogic. If we were to use the rails validation, it would simply append the error resulting in multiple duplicate error messages. I'm not sure if there is a way to override all preexisting errors in rails, which would get rid of the duplicate error message issue.

**Comment by SidharthBansal** [comment]
Wait. Can you please check why Travis is breaking?

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -14,7 +14,9 @@ def create
     using_recaptcha = !params[:spamaway] && Rails.env == "production"
     recaptcha = verify_recaptcha(model: @user) if using_recaptcha
     @spamaway = Spamaway.new(spamaway_params) unless using_recaptcha
-    if ((@spamaway&.valid?) || recaptcha) && @user.save
```
Comment (SidharthBansal) [inline app/controllers/users_controller.rb]: I guess o told you to shift this change to a new pr. This is not related to the current issue.
We aim at solving one unit of functionality in one pr

File: app/models/user.rb [hunk 2]
```diff
@@ -3,9 +3,18 @@ def validate(record)
     if User.find_by(username: record.username) && record.openid_identifier.nil?
       record.errors[:base] << 'That username is already taken. If this is your username, you can simply log in to this site.'
     end
+
+    if record.username.blank?
+      record.errors[:username].clear
```
Comment (SidharthBansal) [inline app/models/user.rb]: This will move to another pr

File: test/integration/signup_flow_test.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      spamaway: {
+        follow_instructions: '',
+        statement1: '',
+        statement2: '',
+        statement3: '',
+        statement4: ''
+      }
+    }
+    
+    assert response.body.include? 'Username should use only letters, numbers, spaces, and .-_@+ please.'
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: Can you assert the number of errors equal to two please ?
Similar to other tests too

File: test/integration/signup_flow_test.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    assert !(response.body.include? 'Username is too short (minimum is 3 characters)')
+  end
+
+  test 'username character and length error messages' do
+    post '/register', params: { 
+      user: {
+        username: '^', 
+        email: 'validemail@gmail.com',
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: Just checking are you using already created fixtures?
Only one field should be inaccurate if you are checking single function.

File: test/integration/signup_flow_test.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        statement4: ''
+      }
+    }
+    
+    assert response.body.include? 'Username is too long (maximum is 100 characters)'
+  end
+
+  test 'username character error messages' do
+    post '/register', params: { 
+      user: {
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: Please divide this into errors. 
Change the title of the test. It is ambiguous

File: test/integration/signup_flow_test.rb [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        username: 'xzd^', 
+        email: 'validemail@gmail.com',
+        password: 'validpassword',
+        password_confirmation: 'validpassword',
+      },
+      spamaway: {
+        follow_instructions: '',
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: May be you have check spam errors too.

File: test/integration/signup_flow_test.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        statement4: ''
+      }
+    }
+    
+    assert response.body.include? 'Username should use only letters, numbers, spaces, and .-_@+ please.'
+    assert response.body.include? 'Username is too short (minimum is 3 characters)'
+  end
+
+  test 'custom blank username error messages' do
+    post '/register', params: {
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: Why is it custom? Any reasons for adding this adjective

**Comment by SidharthBansal** [comment]
I have fever. Please ping me once these changes are done. Sorry for delay

File: test/integration/signup_flow_test.rb (hunk 7 above)
Comment (chen-robert) [reply test/integration/signup_flow_test.rb]: You're right, this adjective is redundant. I will remove this now.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (SidharthBansal) [reply app/controllers/users_controller.rb]: Can you remove this file as it doesn't have any significant changes?

**Comment by chen-robert** [comment]
@SidharthBansal this should be ready to review now, I've implemented all the changes you've requested. Please take a look when you feel better, thanks! 

:tada:

**Comment by SidharthBansal** [comment]
Testing are breaking 
Also @chen-robert 'Partially fixes #6303' phrase in the beginning of the pr will lead to closing of the respective issue on merge. Reason: regex /[Ff]ixes #d+/ is checked 
So please write part of .....

**Comment by SidharthBansal** [comment]
Tests* are breaking

**Comment by Uzay-G** [comment]
Hey @SidharthBansal the tests aren't breaking because of @chen-robert's changes but because of this bug https://github.com/publiclab/plots2/issues/6950

File: app/models/user.rb [hunk 8]
```diff
@@ -6,6 +6,10 @@ def validate(record)
   end
 end
 
+# Fixes authlogic's broken regex.
+# See https://github.com/publiclab/plots2/pull/6932#issuecomment-563878155
```
Comment (SidharthBansal) [inline app/models/user.rb]: Not an ideal way to write comments.

File: test/integration/signup_flow_test.rb [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+          statement1: I18n.t('spamaway.human.statement1'),
+          statement2: I18n.t('spamaway.human.statement1'),
+          statement3: I18n.t('spamaway.human.statement1'),
+          statement4: I18n.t('spamaway.human.statement1')
+        }
+      }
+      
+      assert response.body.include? '2 errors prohibited this user from being saved'
+      assert response.body.include? 'Username should use only letters, numbers, spaces, and .-_@+ please.'
+      assert response.body.include? 'Username is invalid'
```
Comment (SidharthBansal) [inline test/integration/signup_flow_test.rb]: Just checking here, Username is invalid 
should be removed as per your initial proposed changes. Where is that PR?
I want to merge it quickly

**Comment by SidharthBansal** [comment]
I am not able to find your other prs which I told you. Can you please open them and ping me up?

**Comment by chen-robert** [comment]
@SidharthBansal can we merge the PRs one at a time? This way we can avoid the risk of merge conflicts. Once we get this merged, I'll make a PR for email validation and `Invalid username` duplicate errors.

File: app/models/user.rb (hunk 8 above)
Comment (SidharthBansal) [reply app/models/user.rb]: >  See https://github.com/publiclab/plots2/pull/6932#issuecomment-563878155

This is not the ideal way to write the comments

Comment (SidharthBansal) [reply app/models/user.rb]: This is

**Comment by SidharthBansal** [comment]
1. Invalid username
2. Invalid email
3. username other tests
4. email other tests
this way I will proceed.
You will have merge conflicts for sure. So, please create the prs. We can set up a time and do merging and conflict resolution side by side in a single shot
Either in your early morning or in your night, we both will be available to communicate. I will prefer your early morning.
thanks

**Comment by chen-robert** [comment]
@SidharthBansal 

Invalid email: https://github.com/publiclab/plots2/pull/6965
Invalid username: https://github.com/publiclab/plots2/pull/6964
Email tests: https://github.com/publiclab/plots2/pull/6966

Username tests are included in this PR. 

Would Sunday morning PST work for you? Thanks

**Comment by chen-robert** [comment]
Recaptcha error: https://github.com/publiclab/plots2/pull/6967

**Comment by SidharthBansal** [comment]
If your prs are completed then I will review and merge it now

**Comment by SidharthBansal** [comment]
Anything remaining?

**Comment by SidharthBansal** [comment]
Hi can you please write the order in which prs needs to be merged?

**Comment by chen-robert** [comment]
@SidharthBansal They can be merged in any order. If there is a merge conflict, let me know and I will work to get it resolved.

**Comment by chen-robert** [comment]
After this gets merged, I will create FTOs for the "Email should look like an email address." and "2 errors prohibited this user from being saved" you described. 

Thanks!

**Comment by SidharthBansal** [comment]
Cool

**Comment by SidharthBansal** [comment]
Superb
